### PR TITLE
[CALCITE-7696] COVAR result type is derived only from first argument type

### PR DESCRIPTION
**Pull request by mihaibudiu**
## Jira Link

[CALCITE-7696](https://issues.apache.org/jira/browse/CALCITE-7696)

## Changes Proposed

The result type is now defined to be the least restrictive of the two input types.

One additional related bug was fixed: the division required by these functions in the last step also needs to also use a wider type.

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 1]
```diff
@@ -16159,6 +16159,9 @@ void testTimestampDiff(boolean coercionEnabled) {
     f.checkType("covar_pop(CAST(NULL AS INTEGER),CAST(NULL AS INTEGER))",
         "INTEGER");
     f.checkAggType("covar_pop(1.5, 2.5)", "DECIMAL(2, 1) NOT NULL");
+    // [CALCITE-7696] Result type is the least restrictive of the two
+    // argument types, not the type of the first argument
+    f.checkAggType("covar_pop(1, cast(2 as double))", "DOUBLE NOT NULL");
```
Comment (xuzifu666) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16164]: Right now only the INT + DOUBLE case is covered. It would be safer to add a few more edge cases:
1. Mixed decimal/numeric combinations, e.g. DECIMAL + DOUBLE, DECIMAL + INT
2. Nullable scenarios: both arguments NULL, or one of them NULL

**Review by xuzifu666 — APPROVED** [review]
Overall LGTM, only left a comment for test case.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16164]: Added more tests, but by mistake I amended my commit, I apologize

Comment (xuzifu666) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:16164]: I think these changes are fine.
